### ISSUE TITLE
I meet a problem when I runs angr to get all paths and their correpoinding test inputs for a given function

### ISSUE THREAD

**Issue by brokendragon** [angr/angr#103]
Hi

 I want to explare all paths of a given function, and then get all cocrete inputs for the paths. I find the caller object can find all paths, so I type commands as following.
```
import angr
def main():
    b = angr.Project("symbolic_pointer")
    cfg = b.analyses.CFG(keep_state=True)
    target_func = cfg.kb.functions.function(name="test")
    print target_func
    p = b.factory.path()
    x = p.state.memory.load(0x1000, 4)
    y = p.state.memory.load(0x2000, 4)
    c = b.surveyors.Caller(0x4005a6, (0x1000, 0x2000), start=p)
    print tuple(c.iter_returns())
    print c
    print c.found[1]
    state = c.found[3].state
    solution = state.se.any_int(state.memory.load(0x1000, 4))
    print solution
res = main()
```
The source of tested code is
```
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>

int test(int* x, int *y)
{
  //int y = atoi(x);
  int a[4] = {0, 1, 2, 3};
  switch (a[*x - 10]) {
  case 0:
    printf("index is 0\n");
    return 0;
  case 1:
    printf("index is 1\n");
    return 1;
  case 2:
    printf("index is 2\n");
    return 2;
  case 3:
    printf("index is 3\n");
    return 3;
  default:
    return 4; 
  } 
}
void main()
{  
  int x = 0, y = 0;
  test(&x, &y);
}
```

However, I got the same value of varaible x for all paths. Below is the testing result.
```
Function test [0x4005a6]
  Syscall: False
  SP difference: 0
  Has return: True
  Returning: True
  Arguments: reg: [72], stack: [8L, 0L]
  Blocks: [0x400660, 0x400601, 0x400623, 0x4005a6, 0x400606, 0x400651, 0x400608, 0x40062a, 0x40064c, 0x400645, 0x4005f1, 0x400612, 0x400634, 0x4005f6, 0x400619, 0x4005fa, 0x40063b, 0x4005fc, 0x400665]
  Calling convention: System V AMD64 - AMD64 [<rdi>, [8h]]
((<BV64 0x4>, <Path with 6 runs (at 0x4004b0)>), (<BV64 0x1>, <Path with 7 runs (at 0x4004b0)>), (<BV64 0x4>, <Path with 7 runs (at 0x4004b0)>), (<BV64 0x3>, <Path with 10 runs (at 0x4004b0)>), (<BV64 0x2>, <Path with 9 runs (at 0x4004b0)>), (<BV64 0x0>, <Path with 9 runs (at 0x4004b0)>))
<Explorer with paths: 0 active, 0 spilled, 0 deadended, 0 errored, 0 unconstrained, 6 found, 0 avoided, 0 deviating, 0 looping, 0 lost>
<Path with 7 runs (at 0x4004b0)>
167772288
```

That is to say, I got the same value 167772288 no matter I change state = c.found[3].state to state = c.found[0].state, state = c.found[1].state or any other paths.

Am I using Angr in an incorrect way? Or the caller object is not capable of getting concrete test inputs?

Regards

Ting Chen

**Comment by brokendragon** [comment]
I seems that Angr does not get correct concrete inputs when it encounters symbolic pointers. I change the source code `switch (a[*x - 10])` into `switch (*x - 10)`, I can get right inputs. But if my assumption holds, why Angr can find all feasible paths correctly?

**Comment by ltfish** [comment]
In your original example, the input variable `x` of function `test` is `0x1000`, which should be an uninitialized memory cell (I didn't see you initializing it anywhere in your script). That means `*x` is fully unconstrained (can be any value), and `&a[*x - 10]` could point anywhere in the memory. If `&a[*x - 10]` points to an uninitialized memory cell, then `a[*x - 10]` is again a fully unconstrained symbolic value, which will be constrained to different values (like `0`, `1`, `2`, `3`) based on different paths that angr executes afterwards. However, `a[*x - 10] == 0` does not necessarily mean `*x == 10`, since it still holds when `&a[*x - 10]` points to an unconstrained memory cell. Therefore angr is giving you the expected result.

**Comment by brokendragon** [comment]
@ltfish I am not clear about the term 'initialized'. Do you mean the memory cells that are defined explicitly in my test script or the source code? The memory cells such as stack are not considered as initialized, for example
```
void fun()
{
  int a[2] = {0}
}
```
Then, `a[-1]` and `a[3]` (may fail to be compiled, but I can use pointer computation to pass compiling) should point to somewhere in the stack. If I run `fun`, I can get concrete values of `a[-1]` and `a[3]` anyway. In this example, does Angr consider `a[-1]` and `a[3]` are uninitialized, so `a[-1]` and `a[3]`  can be assigned to random values?

**Comment by rhelmot** [comment]
It's a little hard to follow your example, but you have to remember that C code is, in the end, just an abstraction over assembly code, and angr's analysis runs on assembly code. "uninitialized" means that there is no known value for the memory when you read from it. By default in angr, all reads from uninitialized values return unconstrained symbolic variables.

So, if you have a pointer to an array and you use it to access a[-1], the result will depend on the memory layout of the program. If something else is using the memory pointed to by &a[-1], then the read will return that value, or if a previous function call used that region of memory as part of its stack frame, then you'll get back whatever value was written there previously. However, if nothing has ever been written to that address before, you'll get back a symbolic value. This makes sense because when you're analyzing a program with incomplete state, like in your example where you start execution from the start of the function, it's possible that the value at that address could be user-controlled from a previous function call.

**Comment by ltfish** [comment]
@rhelmot's answer is 100% correct.

**Comment by F4lc0nCt** [comment]
Hi,
I would like to discuss a bit more about the access to uninitialised parts of memory through an array. In the case of @brokendragon, there was a constraint with the `switch(a[*x-10])` and I understood how it can affect the symbolic variables. However, I have encountered a case where I was calling a value in an array like `array[i]` and Angr generated some constraints with this instruction. I do not understand how it can generate new constraints because executing `array[i]`is just a reading instruction of the memory.

To be more accurate in my description, I have written an example:
Let's take the following code called `array_ptr.cpp`:
```
#include <iostream>
#include <stdlib.h>     /* atoi */

int arr[10]={0,1,2,3,4,5,6,7,8,9};

int main(int argc, char **argv)
{
        int i = atoi(argv[1]); 
	return arr[i];
}
```
It is a simple code with no conditional jump.
After generating the CFG, I was able to deduce that the final node was at the address:`0x80485cf`
Thus, in order to analyse my binary, I used this Python code:
```
import angr, claripy

path2Bin = "myDir/array_ptr"
b = angr.Project(path2Bin, load_options={'auto_load_libs':False})
argv = [b.filename]
sym_arg_size = 4 #4 Bytes for one int
sym_i = claripy.BVS('sym_i',8*sym_arg_size)
argv.append(sym_i)

#Output of the graph
target=0x80485cf

#Initialisation of entry state
etat = b.factory.entry_state(args=argv)

#Constraint on input format, i has the format "XXXX" with X\in{0, 1, ..., 9}
for byt in range(0,4):	
	etat.se.add(argv[1][(byt*8+7):(byt*8)]>=0x30)
	etat.se.add(argv[1][(byt*8+7):(byt*8)]<=0x39)

#Create path group and launch exploration
path_group=b.factory.path_group(etat)
path_group.explore(find=target)
pfo=path_group.found[0]
pfo_next = pfo.step()[0]
sol_before=pfo.state.se.any_n_str(sym_i,5)
sol_after=pfo_next.state.se.any_n_str(sym_i,5)

#Print solutions
print "Nb of Sol Before the Step: "+str(len(sol_before))
for elt in sol_before:
    print elt
print "Nb of Sol After the Step: "+str(len(sol_after))
for elt in sol_after:
    print elt
```
I obtained the following outputs:
```
Nb of Sol Before the Step: 5
0000
0108
0109
0908
0909
Nb of Sol After the Step: 1
0000
```
Thus, accessing the memory at a symbolic location has led to a new constraints set. I have run more complex examples where the symbolic index was stuck to "9999" or "0000". If I add an index check condition as `i<10` the problem disappears. However, I would like to know why Angr generates these new constraints.
I have tried to check the Guards of the instructions and to set breakpoints before the event "add constraint" but I was unable to detect the update of the constraints set. I have also tried to analyse the execution stack of the `step()` method in the debug mode in Python but it was to difficult due to the number of steps.
Thank you.

**Comment by rhelmot** [comment]
Yes, the behavior you describe is correct. Accessing memory at a symbolic address is a very hard problem to do correctly, and so one of the things that can happen when you perform a "symbolic read", as we call it, is that we concreteize the read address.

**Comment by F4lc0nCt** [comment]
I have just run some examples with some additional conditions and I was a bit confused because Angr was able to cope with small overheads. For instance, if I add a guard `i<15` or `i<99` (thus, it could lead to an outside reading of the array definition), then Angr will not add new constraints during the access to the array whereas for `i<260`, I will have new constraints and the solver will give me `i="0000"`. Is it due to the difficulty of doing correctly "symbolic reading"?

**Comment by ltfish** [comment]
> It seems that the critical value is 256.

When `max(read_addr) - min(read_addr)` is less than or equal to a predefined range, angr will not concretize the read address. This range is controlled by `SimMemory._read_address_range`, which is 1024 by default. That explains why the critical value is 256 - `256 * sizeof(int) = 1024`.

You may want to take a look at how angr concretizes memory read/write addresses in `SimSymbolicMemory`. `concretize_read_addr()` is a good start :-)

**Comment by ltfish** [comment]
Oh one more thing, you can definitely adjust `SimMemory._read_address_range` based on your needs.

**Comment by F4lc0nCt** [comment]
Thank you very much. By giving me the locations of the "concretization" operations, I was able to study the behaviour of Angr with read/write instructions and to understand it.
By the way, based on what I have used so far, I am quite impressed by the possibilities of Angr.

**Comment by ltfish** [comment]
> By the way, based on what I have used so far, I am quite impressed by the possibilities of Angr.

We are glad you say so!

One quick note: it's `angr`, not `Angr` ;)

**Comment by Chozmao** [comment]
hi, i am a beginner and just find this issue. i am facing the same problem. How can i initialize the value?
i have tried something like this `a = s.se.BVS("some_name", 32)` then `s.memory.store(0x1000, a)`. 
i cannot get the correct value of input x for different path. how can i get the right generated test cases by using symbolic execution.
thanks

**Comment by ltfish** [comment]
@Chozmao please fully describe your question in a new issue. I'm seeing this issue as resolved for now.
